2.0-proposal/readme.md

```diff
@@ -0,0 +1,183 @@
+
+### Our internal state has changed! [internal]
+
+The internal state of the unit has changed and we need to react to that.
+
+Ex: A file changing (`file.changed.<xyz>`), landscape messing about on the system (removal of `apt.installed.xyz`), a flag being toggled or the data of a flag being changed (`config.changed` unsets `config.checked`).
+
+<!-- Actions is a tricky one. I'd wait a bit to include actions in the story until we're more certain of where we want to go.
+
+### Action requested! [external]
+
+The operator has requested an action that shouldn't significantly impact the state of the application.
+
+Ex: `smoke-test`, `run-job`, `create-backup`, ... -->
+
+# When do triggers and handlers run?
+
+The flow of the framework is the following:
+
+1. The reactive framework will run a handler when all its preconditions are met. A handler that has run is inactive. It will not run again until one of the flags it reacts to has changed. This change can be either the flag being toggled, or the flag data changing.
+
+2. After each handler run, all internal events will be processed. The reactive framework will not remove flags itself. Only triggers and handlers can do that.
+
+<!-- We should do step 2. recursively. This makes it possible for charmers to create deadlocks/ infinite loops. It would be good if the reactive framework could notice an infinite loop and crash. -->
+
+<!-- Triggers or "Connected States" fix a number of issues:  -->
+<!-- multiple handlers can watch the same file for changes by creating a trigger on `file.changed.xyz` that flags `myhandler.file.changed.xyz` https://github.com/juju-solutions/charms.reactive/issues/25 https://github.com/juju-solutions/charms.reactive/issues/44 -->
+<!-- when_resource_changed is a trigger that flags `resource.changed.xyz` https://github.com/juju-solutions/charms.reactive/issues/87 -->
+
+
+3. After all triggers have been processed, the reactive framework will run the next handler whose preconditions are met. (back to 1.)
+
+The reactive framework goes into sleep mode when there isn't a handler to execute anymore at step 1. The reactive framework wakes up when an event happens.
+
+# What is the difference between external and internal events?
+
+Handlers run transactionally. Internal events are processed after a handler run. External events are processed when the reactive framework wakes up.
+
+ - **All** internal events are processed after **every** handler run.
+ - **Only one** external event is processed when the reactive framework wakes up. If there are still external events waiting to be processed at the moment the reactive framework goes to sleep, it will wake back up immediately and process the next external event.
+
+# Show me the code!
+
+
+```python
+# Register triggers
+charms.reactive.register_trigger(
+  event='config.changed',
+  unset='config.checked')
```

Comment:
Ok, so sorry for the continued push back on this, but I'm still a having a hard time accepting the additional concept of triggers into the library when it can perhaps be achieved in different ways.

As I understand it, essentially this is something (the trigger) that will run *before* any handlers so that flags can be twiddled before handlers conditions' are evaluated.

Conceptually, this is very similar to using a preflight type decorator.  To avoid confusing those two concepts:

```python
@run_before_any_handlers
def twiddle_config():
    if flag('config.changed'):
        set_flag('mylayer.config.changed')
    else:
        clear_flag('mylayer.config.changed')  # just to make sure!
```

The key here, for me, is to try to keep things simple, but also enable resolving a bit of the _can't determine the order that things are run in_ issue.  The syntax above is similar to preconditions and allows you to do more things than a trigger would (although, conceptually, you could make the trigger call a function).  In fact, the trigger could be re-written as:

```python
@trigger(event='config.changed')
def the_config_has_changed():
    set_flag('mylayer.config.changed')
```

Custom events, and triggers, seems to add _more_ to the charms.reactive library, but one of the goals is to do less.